### PR TITLE
Add Kubelet tmpfs storage

### PR DESCRIPTION
**Pull request by pmorie**
Needed for #4602; split out from #4625.

### CODE REVIEW

**Comment by pmorie** [comment]
@thockin @erictune @smarterclayton @markturansky

File: pkg/kubelet/tmpfs_linux.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	if !selinux.SelinuxEnabled() {
+		return defaultMountOptions, nil
+	}
+
+	rootContext, err := selinux.Getfilecon(kl.getRootDir())
+	if err != nil {
+		return "", err
+	}
+
+	// TODO: check out unexpected null char (or something)
```
Comment (pmorie) [inline pkg/kubelet/tmpfs_linux.go]: @mrunalp I got an extra char in the result of `Getfilecon`; know anything about that?

Comment (thockin) [inline pkg/kubelet/tmpfs_linux.go]: Will look when I can.  On a plane now (short layover).  I just hacked up a
tmpfs volume plugin and volume plugin wrappers.  Will send those for
preview when I get to my final destination.

Comment (mrunalp) [inline pkg/kubelet/tmpfs_linux.go]: @pmorie Could be a bug in the API. I will look into it (unless @thockin reports a fix before me) :)
@thockin Looking forward to your tmpfs volume plugin.

Comment (mrunalp) [inline pkg/kubelet/tmpfs_linux.go]: @pmorie @thockin Posted a PR https://github.com/docker/libcontainer/pull/436

The \0 was carried on up with the C string returned by the underlying system call.

**Comment by pmorie** [comment]
@thockin @erictune I think this PR is ready for review; it seems like we might potentially need both this and #5166, since pod volumes do not capture all the use-cases for secrets.    For example, if the kubelet continues to do docker pull (instead of delegating to something else and mounting a shared image store), we might use this to implement #499.

**Comment by alex-mohr** [comment]
@erictune I think you're close to this?  If not, please find someone else to review.

File: cmd/kubelet/app/server.go [hunk 2]
```diff
@@ -130,6 +132,7 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.MasterServiceNamespace, "master_service_namespace", s.MasterServiceNamespace, "The namespace from which the kubernetes master services should be injected into pods")
 	fs.Var(&s.ClusterDNS, "cluster_dns", "IP address for a cluster DNS server.  If set, kubelet will configure all containers to use this for DNS resolution in addition to the host's DNS servers")
 	fs.BoolVar(&s.ReallyCrashForTesting, "really_crash_for_testing", s.ReallyCrashForTesting, "If true, crash with panics more often.")
+	fs.BoolVar(&s.MountTmpfs, "mount_tmpfs", s.MountTmpfs, "If true, kubelet will mount tmpfs storage")
```
Comment (vmarmol) [inline cmd/kubelet/app/server.go]: Mount tmpfs storage sounds slightly vague. Can we clarify what the storage is for?

**Comment by vmarmol** [comment]
I apologize if I rehash any arguments as I get up to speed :) Feel free to tell me if that's the case.

I think it makes sense to start with @thockin's PR #5166 for the per-pod storage. I agree that we need this PR for the Kubelet-wide storage. Maybe we morph it into only that? I am heavily on the per-pod tmpfs camp due to memory usage accounting.

What do we plan to store in the kubelet-wide tmpfs? (You mentioned #499, anything else we have in mind? The Kubelet checkpoint, if it ever comes, is another possibility) The secrets would go into the per-pod volumes no?

**Comment by pmorie** [comment]
@vmarmol Sorry for the late response; I am out of the office this and next week.

Originally @erictune, @thockin and I had discussed a global tmpfs for use
with secrets for an initial path toward storing secrets in tmpfs (which was
something we agreed to in the secrets proposal) - that discussion is in #4602.
I had prototyped the global storage together in #4625, from which I split out
this PR and #5163.  Later @thockin started his work with the plugin wrapper
(which we had discussed for @markturansky's persistent volume work).  Hopefully
that clarifies some of the history.

Since we now have #5166 and #5167, we don't really need this for secrets anymore.
We should probably reasses whether we want the global tmpfs at all for now;
some other use-cases I had in mind were for secrets used by the kubelet on behalf
of pods (ie, secrets associated with the pod's service account but not referenced
via a volume).  All that said, I don't think the semantics of exactly how service
accounts and secrets are going to work together and what the user experience around
them is going to be.  We can probably live without this for now, unless someone
is actively working on those things and needs a global area.

Any opinions, @erictune @thockin @liggitt @smarterclayton @EricMountain-1A ?

**Comment by vmarmol** [comment]
Closing this since #5166 went in.

**Pull request closed without merging**